sklearn/decomposition/_nmf.py

```diff
@@ -795,35 +896,39 @@ def _fit_multiplicative_update(X, W, H, beta_loss='frobenius',
     previous_error = error_at_init
 
     H_sum, HHt, XHt = None, None, None
+
     for n_iter in range(1, max_iter + 1):
-        # update W
-        # H_sum, HHt and XHt are saved and reused if not update_H
-        delta_W, H_sum, HHt, XHt = _multiplicative_update_w(
-            X, W, H, beta_loss, l1_reg_W, l2_reg_W, gamma,
-            H_sum, HHt, XHt, update_H)
-        W *= delta_W
-
-        # necessary for stability with beta_loss < 1
-        if beta_loss < 1:
-            W[W < np.finfo(np.float64).eps] = 0.
-
-        # update H
-        if update_H:
-            delta_H = _multiplicative_update_h(X, W, H, beta_loss, l1_reg_H,
-                                               l2_reg_H, gamma)
-            H *= delta_H
+        for i, slice in enumerate(gen_batches(n=n_samples,
+                                              batch_size=batch_size)):
+            # update W
+            # H_sum, HHt and XHt are saved and reused if not update_H
+            delta_W, H_sum, HHt, XHt = _multiplicative_update_w(
+                X[slice], W[slice], H, beta_loss, l1_reg_W, l2_reg_W,
+                gamma, H_sum, HHt, XHt, update_H)
+            W[slice] *= delta_W
+            # necessary for stability with beta_loss < 1
+            if beta_loss < 1:
+                W[slice][W[slice] < np.finfo(np.float64).eps] = 0.
 
-            # These values will be recomputed since H changed
-            H_sum, HHt, XHt = None, None, None
+            # update H
+            if update_H:
+                delta_H, A, B = _multiplicative_update_h(
+                    X[slice], W[slice], H, A, B, beta_loss,
+                    l1_reg_H, l2_reg_H, i, gamma, rho)
+                H *= delta_H
 
-            # necessary for stability with beta_loss < 1
-            if beta_loss <= 1:
-                H[H < np.finfo(np.float64).eps] = 0.
+                # These values will be recomputed since H changed
+                H_sum, HHt, XHt = None, None, None
+
+                # necessary for stability with beta_loss < 1
+                if beta_loss <= 1:
+                    H[H < np.finfo(np.float64).eps] = 0.
+
+        n_iter += i
```

Comment:
I don't understand what this line is meant for ?

Reply:
I'm adding to the number of iterations the iterations done on different slices.
Let's say `max_iter` is 10 and we have 5 batches, we will run through the 5 batches twice to reach the maximum number of iterations. Is that the right approach?

Reply:
It does not do what you expect currently. Try
```
for i in range(10):
    print(i)
    i += 5
``` 
Also in that case it should be += 1 because right know you're adding 1 for the first batch then 2 for the second, etc..

But I think we should keep the meaning of max_iter to be the number of iterations over the whole dataset. It's what's done for other online estimators. We could still keep track of the number of iterations over minibatches and store it as another attribute.

Reply:
> It does not do what you expect currently

I think it does: the increment is outside the minibatch loop.

> But I think we should keep the meaning of max_iter to be the number of iterations over the whole dataset. It's what's done for other online estimators. We could still keep track of the number of iterations over minibatches and store it as another attribute.

I will add a `iter_offset` parameter like in online dictionary learning https://github.com/scikit-learn/scikit-learn/blob/160debe468890e1dd90b610d5505eb118481cbcf/sklearn/decomposition/_dict_learning.py#L639
Is that ok?

Reply:
I agree that `iter_offset` seems better. If the algorithm needs to know the number of updates performed, it's not enough to know the number of batches seen and the number of updates per batch, because the last batch may be smaller.

Reply:
`iter_offset_` is there. Not sure if I'm using it in the right way...

Reply:
Iḿ resolving this comment as the code has significantly changed.